### ISSUE TITLE
Update Expressions with add/sub & map/list types

### ISSUE THREAD

**Issue by garretcharp** [tywalch/electrodb#70]
Hello, I see this in the readme:
```
Append/Add/Subtract/Remove updates capabilities
Complex attributes (list, map, set)
```

Out of curiosity do you have any ETA for these? Would be very useful to have, I tried to go through and see if I could add them myself but it was a bit hard to follow the code. If you are curious number 1 priority IMO is add/subtract because the any type works fine for now.

Thanks

**Comment by garretcharp** [comment]
Also unrelated to this issue specifically but you should add a way to sponsor the project I would for sure do it because so far this has been a very useful package.

**Comment by tywalch** [comment]
I am nearly done with it actually 👍 There was a lot of related code to be cleaned up and refactored to lay the groundwork for other update methods and complex attribute types. 

Add/Subtract/Append/Remove/Delete will be coming early next week, complex types will come sometime afterward. If you're interested, let me know and I'll post more about the upcoming syntax when I'm at my laptop next.

**Comment by tywalch** [comment]
> Also unrelated to this issue specifically but you should add a way to sponsor the project I would for sure do it because so far this has been a very useful package.

I will look into this 👍

**Comment by garretcharp** [comment]
> I am nearly done with it actually 👍 There was a lot of related code to be cleaned up and refactored to lay the groundwork for other update methods and complex attribute types.
> 
> Add/Subtract/Append/Remove/Delete will be coming early next week, complex types will come sometime afterward. If you're interested, let me know and I'll post more about the upcoming syntax when I'm at my laptop next.

I would appreciate that if you dont mind.

**Comment by tywalch** [comment]
I am now in the stage of writing tests (a lot of ground to cover), but here is the beginnings of those tests so you can see the methods in action:
https://github.com/tywalch/electrodb/blob/improvement/chain-attribute-abstraction-improvements/test/connected.update.spec.js

Basically when this branch is merged there will be 8 methods added to the `update` chain:
method | description
:------  | -----------
`set` | existing method for setting attributes
`add` | new method to add number attributes 
`subtract` | new method to subtract number attributes
`append` | new method to append list/any attributes
`remove` | new method to remove attributes (takes an array)
`delete` | new method to delete set/any attributes
`del` | alias for `delete`
`data` | callback oriented method, similar to where.

The idea with these methods is that _after_ the `update` or `patch` methods, you can supply any number or combination of the methods above (including the same method multiple times), to build you update expression.

**Comment by garretcharp** [comment]
That looks good, happy to have that coming soon. Looking through the tests it doesn't seem you have any testing the "data" method could you explain more what that is for?

**Comment by tywalch** [comment]
The `data()` method will allow for improved type support and make some dynamic queries easier to perform. I pushed some tests to the feature branch since your last message.

```typescript
await entity.update({id})
    .data((attr, op) => {
        // attr - all attributes
        // op - set, add, subtract, append, remove, delete, del
        // unlike `where` does not require formatting a string
        op.add(attr.numProp, 5); 
    })
    .go()
```  

Where this will be a big game changer is with the attribute types `set`, `map`, `list`, and `any`. The `where()` method already allowed for complex types, but the `data()` method will also support complex types. What the means is, lets say you have a property like this:

```
{
    // ... (other properties not shown)
    listProp: [
        {
            prop1: "abc",
            prop2: 10
        },{
            prop1: "def",
            prop2: 15
        }    
    }
}
```

If you wanted to increment `prop2` on the second element, you could use either of the following syntax:

```typescript
await entity.update({id})
    // will be type safe with `list`, `map`
    // can be used with attribute `any` in the mean time
    .data( ({ listProp }, { add }) => {
        add(listProp[1].prop2, 1)
    })
    .go()
```  

```typescript
await entity.update({id})
    // will be possible but won't be type safe
    .add({"listProp[1].prop2", 1})
    .go()
```

**Comment by tywalch** [comment]
As an update, I definitely underestimated some of the nuance of this change. I am wrapping it up, but I still need to add further tests to be sure it'd be a solid release. Thanks for being patient 👍

**Comment by garretcharp** [comment]
> As an update, I definitely underestimated some of the nuance of this change. I am wrapping it up, but I still need to add further tests to be sure it'd be a solid release. Thanks for being patient 👍

No rush at all totally understandable. Thanks for the work you put into this 😄

**Comment by tywalch** [comment]
Feel free to make another PR for map/list/set but the new update methods are now released under `1.2.0` 👍
